pkg/metallb/frr.go

```diff
@@ -0,0 +1,589 @@
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors[neighborIndex].EBGPMultiHop = true
+
+	return builder
+}
+
+// WithHoldTime defines the holdTime placed in the FrrConfiguration spec.
+func (builder *FrrConfigurationBuilder) WithHoldTime(holdTime metav1.Duration, routerIndex,
+	neighborIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Adding a neighbor specific holdTime in the FrrConfiguration %s with holdTime: %v",
+		builder.Definition.Name, holdTime)
+
+	duration := holdTime.Duration
+
+	if duration < time.Second || duration > 65535*time.Second {
+		glog.V(100).Infof("A valid holdtime is between 1-65535")
+
+		builder.errorMsg = "frrConfiguration 'holdtime' value is not valid"
+	}
+
+	// Check if the routerIndex is within bounds
+	if routerIndex < 0 || routerIndex >= len(builder.Definition.Spec.BGP.Routers) {
+		glog.V(100).Infof("Invalid routerIndex: %d", routerIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid routerIndex: %d", routerIndex)
+	}
+
+	// Check if the neighborIndex is within bounds
+	if neighborIndex < 0 || neighborIndex >= len(builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors) {
+		glog.V(100).Infof("Invalid neighborIndex: %d", neighborIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid neighborIndex: %d", neighborIndex)
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors[neighborIndex].HoldTime = &holdTime
+
+	return builder
+}
+
+// WithKeepalive defines the keepAliveTime placed in the FrrConfiguration spec.
+func (builder *FrrConfigurationBuilder) WithKeepalive(keepAlive metav1.Duration, routerIndex,
+	neighborIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Add a keepAlive timer to a specific neighbor in FrrConfiguration %s with keepalive: %v",
+		builder.Definition.Name, keepAlive)
+
+	duration := keepAlive.Duration
+
+	if duration < time.Second || duration > 65535*time.Second {
+		glog.V(100).Infof("A valid keepAlive time is between 1-65535")
+
+		builder.errorMsg = "frrConfiguration 'keepAlive' value is not valid"
+	}
+
+	// Check if the routerIndex is within bounds
+	if routerIndex < 0 || routerIndex >= len(builder.Definition.Spec.BGP.Routers) {
+		glog.V(100).Infof("Invalid routerIndex: %d", routerIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid routerIndex: %d", routerIndex)
+	}
+
+	// Check if the neighborIndex is within bounds
+	if neighborIndex < 0 || neighborIndex >= len(builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors) {
+		glog.V(100).Infof("Invalid neighborIndex: %d", neighborIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid neighborIndex: %d", neighborIndex)
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
```

Comment:
This can be removed if we short-circuit return builder.

Reply:
removed